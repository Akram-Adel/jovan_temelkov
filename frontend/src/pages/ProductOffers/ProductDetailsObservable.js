import store from '../../store';

/**
 * @typedef {Object} State
 * @property {number} id
 * @property {Object.<string, string[]>} interestInfo
 * @property {Object.<string, string[]>} productDetails
 */

/**
 * Instead of polluting the redux store with low level details (such as:
 * which product is selected), I chose to create a custom observable
 * class that is responsible of managing the selected-product and notify
 */
class ProductDetailsObservable {
  /** @type {Function[]} - an array of callback functions */
  _observers = [];

  /** @type {State | null} - state of the selected product */
  _state = null;

  /**
   * Calls all observers with the current state
   */
  _notify() {
    this._observers.forEach((observer) => observer(this._state));
  }

  /**
   * Fetch the product details of the given product and loads it into
   * this observable state.\
   * This will notify all observers with the updated state accordingly
   * @param {number} productId - id of the product to load its details
   */
  loadProduct(productId) {
    const storeProduct = store.getState().productOffers.find((p) => p.id === productId);

    this._state = {
      id: productId,
      interestInfo: {
        effective_rate: ['Effective rate', `${storeProduct.effective_interest_rate}% p.a.`],
        nominal_rate: ['Nominal rate', `${storeProduct.nominal_interest_rate}% p.a.`],
        term: ['Term', `${Math.ceil(storeProduct.number_of_terms / 12)} year/s`],
        availability: ['Availability', storeProduct.availability_notice],
        interest_of_early_termination: [
          'Interest in case of early termination',
          storeProduct.early_termination_effective_interest_rate,
        ],
      },
      productDetails: {
        deposit_guarantee: [
          'Deposit guarantee',
          storeProduct.cms_content.deposit_insurance_name.value,
        ],
        currency: ['Currency', storeProduct.currency],
        investment_amount: [
          'Min/Max investment amount',
          `${storeProduct.min_placement_amount} / ${storeProduct.max_placement_amount}`,
        ],
        withholding_tax: ['Withholding tax', String(storeProduct.tax_number_required)],
      },
    };

    this._notify();
  }

  /**
   * @param {(state: State | null) => void} cb - callback function to call when the state changes
   */
  subscribe(cb) {
    /**
     * when subscribing to the update event, immediately call the
     * callback so that the subscriber matches its internal state to the
     * observable state
     */
    cb(this._state);
    this._observers.push(cb);
  }

  /**
   * @param {(state: State | null) => void} cb - callback function to call when the state changes
   */
  unsubscribe(cb) {
    this._observers = this._observers.filter((sub) => sub !== cb);
  }

  /**
   * Removes all observers and reset this observable's state
   */
  cleanup() {
    this._observers = [];
    this._state = null;
  }
}

export default ProductDetailsObservable;
