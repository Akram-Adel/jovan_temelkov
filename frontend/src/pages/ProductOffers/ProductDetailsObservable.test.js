import ProductOffersResponse from '../../../__mocks__/product-offers_response.json';
import store from '../../store';
import initialState from '../../store/state';
import ProductDetailsObservable from './ProductDetailsObservable';

jest.spyOn(store, 'getState').mockImplementation(() => ({
  ...initialState,
  productOffers: ProductOffersResponse,
}));

describe('ProductDetailsObservable', () => {
  it('notifies observers when a new product is loaded', () => {
    const observable = new ProductDetailsObservable();
    const notifySpy = jest.spyOn(observable, '_notify');

    observable.loadProduct('JTB009');

    expect(notifySpy).toHaveBeenCalled();
  });

  it('formats product-offer into state correctly', () => {
    const observable = new ProductDetailsObservable();
    observable.loadProduct('JTB009');

    observable.subscribe((state) => {
      expect(state).toMatchSnapshot();
    });
  });

  it('subscribe and unsubscribe correctly', () => {
    const observable = new ProductDetailsObservable();
    expect(observable._observers).toEqual([]);

    const observer = jest.fn();
    observable.subscribe(observer);
    expect(observable._observers).toEqual([observer]);

    observable.unsubscribe(observer);
    expect(observable._observers).toEqual([]);
  });

  it('cleanup observers and state correctly', () => {
    const observable = new ProductDetailsObservable();

    observable.loadProduct('JTB009');
    observable.subscribe(() => null);
    observable.subscribe(() => null);

    expect(observable._observers.length).not.toEqual(0);
    expect(observable._state).not.toBeNull();
    
    observable.cleanup();
    expect(observable._observers.length).toEqual(0);
    expect(observable._state).toBeNull();
  });
});
