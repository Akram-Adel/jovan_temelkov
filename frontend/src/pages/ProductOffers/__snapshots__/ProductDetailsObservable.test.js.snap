// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailsObservable formats product-offer into state correctly 1`] = `
Object {
  "id": "JTB009",
  "interestInfo": Object {
    "availability": Array [
      "Availability",
      undefined,
    ],
    "effective_rate": Array [
      "Effective rate",
      "1.17% p.a.",
    ],
    "interest_of_early_termination": Array [
      "Interest in case of early termination",
      0,
    ],
    "nominal_rate": Array [
      "Nominal rate",
      "1.2% p.a.",
    ],
    "term": Array [
      "Term",
      "5 year/s",
    ],
  },
  "productDetails": Object {
    "currency": Array [
      "Currency",
      "EUR",
    ],
    "deposit_guarantee": Array [
      "Deposit guarantee",
      "bis 100.000 EUR",
    ],
    "investment_amount": Array [
      "Min/Max investment amount",
      "10000 / 100000",
    ],
    "withholding_tax": Array [
      "Withholding tax",
      "false",
    ],
  },
}
`;
